SCMOD-12257: Correlation Id Support

simranjeetc opened the pull request:
https://portal.digitalsafe.net/browse/SCMOD-12257

Commit:
Creating a library to include dropwizard bundle & sprint boot filter for logging correlation id

Commit:
Updated package name

Commit:
Updated the header key for correlation Id

buildmachine-sou-jenkins2 commented:
CI Build Link:
http://sou-jenkins2.swinfra.net:8080/job/CAFapi/job/CAFapi~caf-logging~SCMOD-12257~CI

Commit:
Changed the scope of servlet jars to provided

Commit:
Added correlation id support for spring boot interceptors

Commit:
Removed dependency on spring-core

Commit:
Changed scope of dropwizard dependencies to provided.

Commit:
Added ITs

simranjeetc commented:
@dermot-hardy 
I have opened this PR for review. Once it is merged, I will update the jar version in the pom files for all the services which are using this library.  Until then, I will leave the others as draft only.

dermot-hardy commented on pom.xml:
```diff
@@ -79,6 +80,13 @@
+                <version>2.0.0-268</version>
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
```
Including this would create a circular dependency.

(And making it higher priority than the log4j-bom is downgrading the log4j version from 2.11.1 to 2.8.2).

dermot-hardy replied:
Removed this in 9a26ccd when I split the correlation projects into separate modules.

dermot-hardy commented on caf-correlation-id/pom.xml:
```diff
@@ -0,0 +1,161 @@
+        <dependency>
+            <groupId>io.dropwizard</groupId>
+            <artifactId>dropwizard-core</artifactId>
+            <scope>provided</scope>
```
I suspect you're setting these as `scope=provided` because this dependency is not required by all potential consumers.  This is a bit of an abuse of the scope, especially since Maven does have support for optional dependencies by adding `<optional>true</optional>`, but really the right way to deal with this is to split the module up into sub-modules.

dermot-hardy replied:
I split the project into separate modules in 9a26ccd to resolve this.

simranjeetc replied:
I have used `scope=provided` for below mentioned reasons -
1. `CorrelationIdFilter` and `CorrelationIdClientFilter`- These filters implement interfaces `javax.servlet.Filter` and `javax.ws.rs.client.ClientRequestFilter` respectively. The implementation jar varies from project to project. Thus, found it better in this case for consumer to own the implementation jar dependency rather than the library. 
2. To avoid a version mismatch on dependencies. The code in this library does not require any specific version of dropwizard or spring and thus leaving it to the consumer to provide the dependent jars via `provided`.

These dependencies are not optional and thus did not use optional scope. I did start without `provided` scope in the beginning but ran into the above mentioned issues. These issues were reported by analyze plugins and were causing the build to fail. 
I see you kept the provided scope for the `CorrelationIdFilter` but did not do so for the `CorrelationIdClientFilter`. If you are in agreement, let me know I will add the same.

dermot-hardy replied:
As discussed, this comment was relating to `dropwizard-core`.  I didn't remove `scope=provided` from `jakarta.servlet-api`.  I wasn't sure about `jakarta.ws.rs-api` so feel free to re-add it to that dependency if it's causing a problem.

dermot-hardy commented on caf-correlation-id/pom.xml:
```diff
@@ -0,0 +1,161 @@
+        <plugins>
+            <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-failsafe-plugin</artifactId>
```
This is required for running integration tests but so far as I can see you only have unit tests.

dermot-hardy replied:
Scrapped this in 9a26ccd.

dermot-hardy commented on caf-correlation-id/src/main/java/com/github/cafapi/http/interceptors/CorrelationIdInterceptor.java:
```diff
@@ -0,0 +1,50 @@
+import org.springframework.web.servlet.ModelAndView;
+import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;
+
+public final class CorrelationIdInterceptor extends HandlerInterceptorAdapter
```
Note that this approach is deprecated:
https://github.com/spring-projects/spring-framework/pull/25147

I think it is fine for now but we may wish to come back to this at some point.

dermot-hardy replied:
Moved this comment to correct line.

dermot-hardy commented on caf-correlation-id/src/test/java/com/filter/http/spring/TestingWebApplicationTests.java:
```diff
@@ -0,0 +1,96 @@
+        //prepare
+        HttpHeaders httpHeaders = new HttpHeaders();
+        httpHeaders.add(headerName, "UUID1");
+        HttpEntity request = new HttpEntity(httpHeaders);
```
This appears to be causing warnings in the build log:
`Some input files use unchecked or unsafe operations.`

dermot-hardy replied:
Fixed in ce9a0cc.

dermot-hardy commented on caf-correlation-id/src/test/java/com/filter/http/dropwizard/CorrelationIdDropwizardTest.java:
```diff
@@ -0,0 +1,98 @@
+        @Override
+        public void initialize(Bootstrap<Configuration> bootstrap)
+        {
+            bootstrap.addBundle(new CorrelationIdBundle());
```
Causing a build output warning.

dermot-hardy replied:
Fixed in ce9a0cc.

dermot-hardy commented on caf-correlation-id/src/main/java/com/github/cafapi/http/filters/correlationid/CorrelationIdConfigurationConstants.java:
```diff
@@ -0,0 +1,22 @@
+public class CorrelationIdConfigurationConstants
+{
+    public static final String headerName = "CAF-Correlation-Id";
+    public static final String mdcKey = "correlationId";
```
These names are not following the standard Java constants convention.

dermot-hardy replied:
Fixed in c51b436.

dermot-hardy reviewed (changes requested):
The `duplicate-finder-maven-plugin` is reporting lots of warnings:

```
2021-02-03T07:06:12.933663288Z [INFO] --- duplicate-finder-maven-plugin:1.3.0:check (deepcheck-dependencies) @ caf-correlation-id ---
2021-02-03T07:06:12.935268289Z [INFO] Checking compile classpath
2021-02-03T07:06:13.731032292Z [INFO] Checking runtime classpath
2021-02-03T07:06:13.735996134Z [INFO] Checking test classpath
...
2021-02-03T07:06:14.695631150Z [WARNING] Found duplicate and different classes in [org.apache.tomcat.embed:tomcat-embed-el:9.0.33, org.glassfish:jakarta.el:3.0.3]:
2021-02-03T07:06:14.695639853Z [WARNING]   javax.el.ArrayELResolver
2021-02-03T07:06:14.695664629Z [WARNING]   javax.el.BeanELResolver
2021-02-03T07:06:14.696733828Z [WARNING]   javax.el.BeanNameELResolver
2021-02-03T07:06:14.696758507Z [WARNING]   javax.el.BeanNameResolver
2021-02-03T07:06:14.696767390Z [WARNING]   javax.el.CompositeELResolver
...
2021-02-03T07:06:14.708697867Z [WARNING]   javax/servlet/http/LocalStrings.properties
```

dermot-hardy committed:
Fix generics issues in tests

dermot-hardy committed:
Split the project into separate modules

dermot-hardy committed:
Updated the copyright headers

dermot-hardy committed:
Updated constants to standard convention

dermot-hardy committed:
Moved the classes to module-aligned packages

dermot-hardy committed:
Whitespace changes: Source --> Format

dermot-hardy committed:
Added missing final keywords

dermot-hardy committed:
Disallow instantiating the constants module

dermot-hardy commented on caf-correlation/caf-correlation-id-spring/src/main/java/com/github/cafapi/correlation/spring/CorrelationIdInterceptor.java:
```diff
@@ -0,0 +1,54 @@
+import org.springframework.web.servlet.ModelAndView;
+import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;
+
+public final class CorrelationIdInterceptor extends HandlerInterceptorAdapter
```
Note that this approach is deprecated:
https://github.com/spring-projects/spring-framework/pull/25147

I think it is fine for now but we may wish to come back to this at some point.

simranjeetc replied:
Sure

dermot-hardy committed:
Updated to avoid spring-tests being the last project

Deployment is disabled for it and the last project is responsible for deploying all of the project's artifacts.

dermot-hardy reviewed (approved):
Looking at these changes now I think it makes more sense to have them to a separate repository. 
 I've created https://github.com/CAFapi/caf-correlation for this.  We can check them in here for now if you want so long as we move them before the next release.  Moving them will involve resetting the version number back from `1.4.0-SNAPSHOT` to `1.0.0-SNAPSHOT` so it probably makes more sense to do it sooner rather than later.

As far as I can see the main thing missing here now is documentation.  These modules require a write-up to explain what they are and how they should be used.

Commit:
Added provided scope rs-api

Commit:
Removed "-id" from module name and artifact ids

Commit:
Moved caf-correlation module to a separate repository. Hence, removing it from here.

simranjeetc commented:
> Looking at these changes now I think it makes more sense to have them to a separate repository.
> I've created https://github.com/CAFapi/caf-correlation for this. We can check them in here for now if you want so long as we move them before the next release. Moving them will involve resetting the version number back from `1.4.0-SNAPSHOT` to `1.0.0-SNAPSHOT` so it probably makes more sense to do it sooner rather than later.
> 
> As far as I can see the main thing missing here now is documentation. These modules require a write-up to explain what they are and how they should be used.

> /home/simran/.jdks/adopt-openjdk-1.8.0_275/bin/java -Dmaven.multiModuleProjectDirectory=/media/D/codebase/CAFapi/caf-correlation -Dmaven.home=/opt/apache-maven-3.6.2 -Dclassworlds.conf=/opt/apache-maven-3.6.2/bin/m2.conf -Dmaven.ext.class.path=/media/D/softwares/jetbrains-tools-installation/apps/IDEA-U/ch-0/202.8194.7/plugins/maven/lib/maven-event-listener.jar -javaagent:/media/D/softwares/jetbrains-tools-installation/apps/IDEA-U/ch-0/202.8194.7/lib/idea_rt.jar=44829:/media/D/softwares/jetbrains-tools-installation/apps/IDEA-U/ch-0/202.8194.7/bin -Dfile.encoding=UTF-8 -classpath /opt/apache-maven-3.6.2/boot/plexus-classworlds-2.6.0.jar org.codehaus.classworlds.Launcher -Didea.version=2020.2.4 -s /opt/apache-maven-3.6.2/conf/settings.xml org.basepom.maven:duplicate-finder-maven-plugin:1.5.0:check
> [INFO] Scanning for projects...
> [INFO] Inspecting build with total of 7 modules...
> [INFO] Installing Nexus Staging features:
> [INFO]   ... total of 7 executions of maven-deploy-plugin replaced with nexus-staging-maven-plugin
> [INFO] ------------------------------------------------------------------------
> [INFO] Reactor Build Order:
> [INFO] 
> [INFO] caf-correlation                                                    [pom]
> [INFO] caf-correlation-constants                                          [jar]
> [INFO] caf-correlation-jaxrs                                              [jar]
> [INFO] caf-correlation-jservlet                                           [jar]
> [INFO] caf-correlation-spring                                             [jar]
> [INFO] caf-correlation-spring-tests                                       [jar]
> [INFO] caf-correlation-dropwizard                                         [jar]
> [INFO] 
> [INFO] -----------< com.github.cafapi.correlation:caf-correlation >------------
> [INFO] Building caf-correlation 1.0.0-SNAPSHOT                            [1/7]
> [INFO] --------------------------------[ pom ]---------------------------------
> [INFO] 
> [INFO] --- duplicate-finder-maven-plugin:1.5.0:check (default-cli) @ caf-correlation ---
> [INFO] Ignoring POM project!
> [INFO] 
> [INFO] ------< com.github.cafapi.correlation:caf-correlation-constants >-------
> [INFO] Building caf-correlation-constants 1.0.0-SNAPSHOT                  [2/7]
> [INFO] --------------------------------[ jar ]---------------------------------
> [INFO] 
> [INFO] --- duplicate-finder-maven-plugin:1.5.0:check (default-cli) @ caf-correlation-constants ---
> [INFO] Checking compile classpath
> [INFO] Checking runtime classpath
> [INFO] Checking test classpath
> [INFO] 
> [INFO] --------< com.github.cafapi.correlation:caf-correlation-jaxrs >---------
> [INFO] Building caf-correlation-jaxrs 1.0.0-SNAPSHOT                      [3/7]
> [INFO] --------------------------------[ jar ]---------------------------------
> Downloading from company-central: https://svsartifactory.swinfra.net/artifactory/company-central/com/github/cafapi/correlation/caf-correlation-constants/1.0.0-SNAPSHOT/maven-metadata.xml
> [INFO] 
> [INFO] --- duplicate-finder-maven-plugin:1.5.0:check (default-cli) @ caf-correlation-jaxrs ---
> [INFO] Checking compile classpath
> [INFO] Checking runtime classpath
> [INFO] Checking test classpath
> [INFO] 
> [INFO] -------< com.github.cafapi.correlation:caf-correlation-jservlet >-------
> [INFO] Building caf-correlation-jservlet 1.0.0-SNAPSHOT                   [4/7]
> [INFO] --------------------------------[ jar ]---------------------------------
> [INFO] 
> [INFO] --- duplicate-finder-maven-plugin:1.5.0:check (default-cli) @ caf-correlation-jservlet ---
> [INFO] Checking compile classpath
> [INFO] Checking runtime classpath
> [INFO] Checking test classpath
> [INFO] 
> [INFO] --------< com.github.cafapi.correlation:caf-correlation-spring >--------
> [INFO] Building caf-correlation-spring 1.0.0-SNAPSHOT                     [5/7]
> [INFO] --------------------------------[ jar ]---------------------------------
> [INFO] 
> [INFO] --- duplicate-finder-maven-plugin:1.5.0:check (default-cli) @ caf-correlation-spring ---
> [INFO] Checking compile classpath
> [INFO] Checking runtime classpath
> [INFO] Checking test classpath
> [INFO] 
> [INFO] -----< com.github.cafapi.correlation:caf-correlation-spring-tests >-----
> [INFO] Building caf-correlation-spring-tests 1.0.0-SNAPSHOT               [6/7]
> [INFO] --------------------------------[ jar ]---------------------------------
> Downloading from company-central: https://svsartifactory.swinfra.net/artifactory/company-central/com/github/cafapi/correlation/caf-correlation-jaxrs/1.0.0-SNAPSHOT/maven-metadata.xml
> Downloading from company-central: https://svsartifactory.swinfra.net/artifactory/company-central/com/github/cafapi/correlation/caf-correlation-spring/1.0.0-SNAPSHOT/maven-metadata.xml
> [INFO] 
> [INFO] --- duplicate-finder-maven-plugin:1.5.0:check (default-cli) @ caf-correlation-spring-tests ---
> [INFO] Checking compile classpath
> [INFO] Checking runtime classpath
> [INFO] Checking test classpath
> [INFO] 
> [INFO] ------< com.github.cafapi.correlation:caf-correlation-dropwizard >------
> [INFO] Building caf-correlation-dropwizard 1.0.0-SNAPSHOT                 [7/7]
> [INFO] --------------------------------[ jar ]---------------------------------
> Downloading from company-central: https://svsartifactory.swinfra.net/artifactory/company-central/com/github/cafapi/correlation/caf-correlation-jservlet/1.0.0-SNAPSHOT/maven-metadata.xml
> [INFO] 
> [INFO] --- duplicate-finder-maven-plugin:1.5.0:check (default-cli) @ caf-correlation-dropwizard ---
> [INFO] Checking compile classpath
> [INFO] Checking runtime classpath
> [INFO] Checking test classpath
> [INFO] ------------------------------------------------------------------------
> [INFO] Reactor Summary for caf-correlation 1.0.0-SNAPSHOT:
> [INFO] 
> [INFO] caf-correlation .................................... SUCCESS [  0.401 s]
> [INFO] caf-correlation-constants .......................... SUCCESS [  0.207 s]
> [INFO] caf-correlation-jaxrs .............................. SUCCESS [  8.673 s]
> [INFO] caf-correlation-jservlet ........................... SUCCESS [  0.060 s]
> [INFO] caf-correlation-spring ............................. SUCCESS [  0.420 s]
> [INFO] caf-correlation-spring-tests ....................... SUCCESS [ 13.949 s]
> [INFO] caf-correlation-dropwizard ......................... SUCCESS [  8.457 s]
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD SUCCESS
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  33.598 s
> [INFO] Finished at: 2021-02-11T10:23:04+05:30
> [INFO] ------------------------------------------------------------------------
> 

Not getting duplicates with latest changes.

simranjeetc commented:
Closing this PR as we moved this code to a new repository https://github.com/CAFapi/caf-correlation/pull/1
